BlogPost:  VTOrc vs Orchestrator (#1149)
vitessio/website: content/en/blog/2022-09-21-vtorc-vs-orchestrator.md

---
author: 'Manan Gupta'
date: 2022-09-21
slug: '2022-09-21-vtorc-vs-orchestrator'

> tags: ['Vitess','Orchestrator','VTOrc']

Reviewer: can add additional tags like mysql cluster management, failover, etc.

Author: Added a couple more.

Reviewer 2: we should add a few more tags. I don't know what popular search terms are, but maybe one or more of "Failover", "Fault tolerance", "Distributed Systems"

Author: Added them all 👍

> title: 'VTOrc Vs Orchestrator'

Reviewer 2: VTOrc: Vitess-native Orchestrator

Reviewer 2: Just a suggestion. If you don't like it we can publish with original title. Doesn't matter too much.

Author: Renamed 👍

description: "The differences between VTOrc and Orchestrator"
---

> There was an idea. An idea to make Vitess self-reliant. An idea to get rid of the friction between Vitess and an external fault-detection-and-repair tool. An idea that gave birth to VTOrc…

Reviewer 3: an external fault-detection-and-repair tool
or 
the external fault-detection-and-repair tools ??

is there any specific tool (orchestrator) or different companies have their own solutions?

Author: You're right, it is supposed to be `and external fault-detection-and-repair tools`

Both [VTOrc](https://vitess.io/docs/user-guides/configuration-basic/vtorc/) and [Orchestrator](https://github.com/openark/orchestrator) are tools for managing MySQL instances. If I were to describe these tools using a metaphor, I would say that they are kinda like the monitor of a class of students. They are responsible for keeping the MySQL instances in check and fixing them up in case they misbehave, just like how a monitor ensures that no mischief happens in the classroom. 

VTOrc started off as a fork of Orchestrator, which was then custom-fitted to the Vitess use-case running as a native Vitess component. VTOrc and Orchestrator are like twins, if you look from far away, you might think that they are the same and do the same thing, but the closer you look, the more differences emerge in architecture, functionality, and flexibility. If you are interested in appreciating the differences between these two twins, then you are at the right place. Let’s jump right into it.

> Most of the differences between the two derive largely from the fact that VTOrc can be opinionated about a bunch of things because it only has to solve one use-case which is the Vitess use-case, but Orchestrator was built to work with virtually any MySQL topology. For example, Vitess currently doesn’t support hierarchical replication. You have 1 primary tablet in a shard and replicas of those primaries. There is no cascading replication. No replicas are replicating from other replicas in steady-state. But Orchestrator allows this configuration. So, VTOrc doesn’t have to worry about the hierarchical topology and can do away with this flexibility that Orchestrator provides in favor of simpler code that ties in with other Vitess components like [vtcltd](https://vitess.io/docs/user-guides/configuration-basic/vtctld/).

Reviewer 3: can do away with this flexibility or you meant
can get away with this flexibility

Author: No, I think we want to keep `do away with this flexibility`. VTOrc is letting go of the flexibility to run with arbitrary MySQL topologies, because it is only concerned with Vitess, which doesn't support them anyways. So we are `doing away` with this flexibility.

# Discovery

> The first point of difference between the two applications is that Orchestrator is a complete tool in itself while VTOrc is part of a much larger framework, Vitess. So VTOrc can rely on other components of Vitess to simplify its design even further. Let us look at MySQL instance discovery to understand this better.

Reviewer 3: so what should we call vtorc and orchestrator .. Are they application , service (micro), tool or component ?

Author: All of the above? It's an application/tool running as a service, also a component of Vitess.

Reviewer 2: or you can delete "applications" and the sentence still makes complete sense.

Author: Removed `applications`

> From the Orchestrator documentation on [Discovery](https://github.com/openark/orchestrator#discovery), `Orchestrator actively crawls through your topologies and maps them. It reads basic MySQL info such as replication status and configuration`. VTOrc here takes a completely different approach. In Vitess, all MySQL instances have a sidecar associated with them called [VTTablets](https://vitess.io/docs/reference/programs/vttablet/) and these VTTablets register themselves in the [topo server](https://vitess.io/docs/concepts/topology-service/). So VTOrc can just go ahead and ask the topo server for the exhaustive list of VTTablets for the shards that it is concerned with and use those records to discover and poll the underlying MySQL instances.

Reviewer 3: [suggestion]
VTOrc here takes a completely different approach
>> On the other hand, VTOrc here takes a completely different approach

Author: Fixed.

> From the Orchestrator documentation on [Discovery](https://github.com/openark/orchestrator#discovery), `Orchestrator actively crawls through your topologies and maps them. It reads basic MySQL info such as replication status and configuration`. On the other hand, VTOrc here takes a completely different approach. In Vitess, all MySQL instances have a sidecar associated with them called [VTTablets](https://vitess.io/docs/reference/programs/vttablet/) and these VTTablets register themselves in the [topo server](https://vitess.io/docs/concepts/topology-service/). So VTOrc can just go ahead and ask the topo server for the exhaustive list of VTTablets for the shards that it is concerned with and use those records to discover and poll the underlying MySQL instances.

Reviewer 2: ```suggestion
From the Orchestrator documentation on [Discovery](https://github.com/openark/orchestrator#discovery), `Orchestrator actively crawls through your topologies and maps them. It reads basic MySQL info such as replication status and configuration`. On the other hand, VTOrc here takes a completely different approach. In Vitess, all MySQL instances have a sidecar associated with them called [VTTablet](https://vitess.io/docs/reference/programs/vttablet/). These VTTablets register themselves in the [topo server](https://vitess.io/docs/concepts/topology-service/). So VTOrc can just go ahead and ask the topo server for the exhaustive list of VTTablets for the shards that it is concerned with and use those records to discover and poll the underlying MySQL instances.
```

Author: Fixed

> What we discussed now was the discovery of the MySQL instances, but there is another dimension to discovery. It is the topology that the MySQL instances are supposed to be running in. From the VTOrc standpoint, we already know Vitess only supports a single hierarchy of MySQL replication with no co-primary scenarios, so there is only supposed to be one primary and all the other MySQL instances replicating from it. As to who the primary is, that information is stored in the topo-server. For Orchestrator, however, there is no central location where the desired topology is stored, it has to infer that based on the current topology configuration as it stands and any changes the user might make going forward. In other words, VTOrc’s topology discovery and maintenance are somewhat declarative in nature whereas Orchestrator works in a more imperative fashion.

Reviewer 2: ```suggestion
There is another dimension to discovery that goes beyond the current state. It is the topology that the MySQL instances are **supposed** to be running in. From the VTOrc standpoint, we already know Vitess only supports a single hierarchy of MySQL replication with no co-primary scenarios, so there is only supposed to be one primary per shard and all other MySQL instances in the shard should be replicating from it. As to who the primary is, that information is stored in the topo server. For Orchestrator, however, there is no central location where the desired topology is stored, it has to infer that based on the current topology configuration as it stands and any changes the user might make going forward. In other words, VTOrc’s topology discovery and maintenance are somewhat declarative in nature whereas Orchestrator works in a more imperative fashion.
```

Author: Fixed

# Synchronization

> Just as the presence of the topo-server greatly simplifies MySQL discovery, it also helps in synchronization. 

Reviewer 2: Some places it is "topo server", some places it is "topo-server". Can you choose one way of saying it? Maybe expand it to "topology server" instead.

Author: Changed them all to `topology server`

Orchestrator provides two ways of providing [high availability](https://github.com/openark/orchestrator/blob/master/docs/high-availability.md#orchestrator-high-availability), the first one using the consensus algorithm Raft, and the other using a shared backing store for the orchestrator nodes.

VTOrc on the other hand relies on the existing functionality of shard locking that Vitess uses for synchronization between various actors. Since the topo-servers are reliable key-value stores, which run some consensus algorithms under the hood, VTOrc can rely on them for this functionality of only allowing a sole actor to acquire a shard lock thereby guaranteeing synchronization. This allows multiple VTOrc instances to monitor the same cluster without knowing or caring about the existence of the others.

# Ephemeral Data 

> In Vitess, the topo-server is responsible for storing persistent and durable data like the topology structure, the durability policy, etc. This allows VTOrc to function with only storing ephemeral data. It doesn’t need any data to be persisted across restarts, which makes VTOrc a truly cloud-native component, since it can be restarted at will and its data gets populated again from the topo-server.

Reviewer 2: cloud-native or stateless?
We may as well add "cloud native" to the tags in the metadata.

Author: I added cloud-native to the tags. 
We can't really say `stateless`, because it isn't. The point is that the state is ephemeral, not that it doesn't have a state.

# UI
Orchestrator comes bundled with an attractive [user interface](https://github.com/openark/orchestrator/blob/master/docs/using-the-web-interface.md) which serves as the final piece of an already impressive tool. It allows the users to look at the current topology structure, investigate the settings of all the MySQL nodes and even change them, right from the UI! 

> In Vitess however, the user interface isn’t just about the running MySQL instances, but should also incorporate other Vitess components like the vttablets, and vtgates. It should allow the users to look at their current Vitess configurations like [VSchema](https://vitess.io/docs/concepts/vschema/) and should also allow them to change it seamlessly. Here yet another Vitess component comes to our aid, [VTAdmin](https://vitess.io/docs/reference/vtadmin/). It is the administration tool offered by Vitess providing an API and a web interface. The team is working towards incorporating all the data and functionality that the VTOrc UI provides (which it inherited from Orchestrator) into VTAdmin, at which point VTOrc’s UI will be deprecated and removed.

Reviewer 2: ```suggestion
In Vitess however, the user interface isn’t just about the running MySQL instances, but should also incorporate other Vitess components like the vttablets, and vtgates. It should allow the users to look at their current Vitess configurations like [VSchema](https://vitess.io/docs/concepts/vschema/) and should also allow them to change it seamlessly. Here yet another Vitess component comes to our aid, [VTAdmin](https://vitess.io/docs/reference/vtadmin/). It is the administration tool offered by Vitess providing an API and a web interface. The team is working towards incorporating all the data and functionality that the VTOrc UI provides (which it inherited from Orchestrator) into VTAdmin, at which point VTOrc’s stand-alone UI will be deprecated and removed.
```

Author: Fixed.

VTOrc, on the other hand, has access to VTTablet metadata along with the MySQL instance, which allows it to infer that a VTTablet is in the process of taking a backup and its replication should not be fixed without explicit action from the VTTablet.

# Future Scope
There are multiple possibilities for VTOrc to go above and beyond in the failure scenarios that it can handle as compared to Orchestrator. VTOrc can handle failures related to the VTTablets as well and not just the MySQL instances. It can subscribe to the VTTablet health checks to accomplish the same. The possibilities are endless and truly exciting.

> With VTOrc going GA soon, it is the perfect time for you to try it out. If you do, please provide us feedback on your experience via GitHub or Slack.

Reviewer 2: This is a good "call to action".
Let's add links to GH new issue template and Slack workspace.

Author: Added the links to both.
